src/java/htsjdk/samtools/util/AsyncBufferedIterator.java

```diff
@@ -129,17 +129,43 @@ public void close() {
             }
         }
     }
+    
+    private void ensureHasNext() {
+        if (!currentBlock.hasNext()) {
+            // Rethrow any exceptions raised on the background thread
+            // at the point the exception would have been encountered
+            // if we had performed synchronous iteration
+            raiseBackgroundThreadException();
```

Comment:
So any exception from the background thread will be thrown only at the **end** of processing the read-ahead block of `T`s, not at the beginning. Is that intended? I'd think that throwing as fast as we know there was a problem would be preferable.